Uberjar run configuration

nullnumber1 opened the pull request:
- Добавлена задача uberjar в deps.edn
- Изменен csa.clj, чтобы корректно исполнялась AOT и не падала от невозможности прочитать файлы конфигурации
- Обновлен Dockerfile, теперь он использует course-bot.jar для запуска приложения.
- Добавлена задача clean в Makefile, которая очищает проект от всех артефактов, за исключением исходных файлов.

Не знаю как оно запустится, т.к без отсутствия конфигов ничего не протестировать. Но по крайней мере всё собирается

ryukzak reviewed (changes requested):
```shell
~/src/course-bot uberjar *2 ?1 ❯ make deploy 
docker run --name csa-bot --restart=always -d -v /Users/penskoi/src/course-bot/../edu-csa-internal:/edu-csa-internal -v /Users/penskoi/src/course-bot/../csa-db:/csa-db csa-bot
bb7b7cf68c8661287366e92ddd373a564d271e44e30ace142c39716029d60a09
~/src/course-bot uberjar *2 ?1 ❯ docker logs csa-bot
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
Error: Could not find or load main class course_bot.csa
Caused by: java.lang.ClassNotFoundException: course_bot.csa
```

For test you can use example config. Of course, you will not connect to telegram, but you can see that other thing is ok.

When finish -- make review request.

nullnumber1 commented on src/course_bot/csa.clj:
```diff
@@ -1,110 +1,114 @@
-          (recur (polling/start token bot-api)))
-      (recur channel)))
-  (println (tr :csa/stop)))
+                     (defn -main [& _args]
```
It actually seems to run with `defn -main`. I think that it's obligatory to specify the entry point of application like that because it will fail otherwise
`➜  course-bot git:(master) ✗ java -jar course-bot.jar
Exception in thread "main" java.lang.UnsupportedOperationException: course-bot.csa/-main not defined
`

nullnumber1 commented:
Updated csa.clj and added (:gen-class) directive to generate a Java class. That's why it was failed to be found previously. Now seems to be working fine
`➜  course-bot git:(master) ✗ make deploy            
docker run --name csa-bot --restart=always -d -v /Users/admin/Documents/Work/course-bot/../edu-csa-internal:/edu-csa-internal -v /Users/admin/Documents/Work/course-bot/../csa-db:/csa-db csa-bot
54f5b541315726a89c3873bd4359445d586bade1b7687b5ad5144ca6b4911722
➜  course-bot git:(master) ✗ docker logs csa-bot
Бот активирован, мой господин!
Mar 27, 2023 9:29:02 PM morse.polling invoke
WARNING: Got error from Telegram API, stopping polling
Mar 27, 2023 9:29:02 PM morse.polling invoke
WARNING: Got error from Telegram API, stopping polling
Mar 27, 2023 9:29:03 PM morse.polling invoke
WARNING: Got error from Telegram API, stopping polling
`

ryukzak commented on src/course_bot/csa.clj:
```diff
@@ -24,87 +25,91 @@
     :unknown-1 "Неизвестное сообщение: %s, а вы точно мой господин?"}}})
 
-(def conf (misc/get-config "../edu-csa-internal"))
+(def conf (delay (misc/get-config "../edu-csa-internal")))
```
Why we need this instead the normal variables?

nullnumber1 replied:
because uberjar build failed for me before I converted everything related to configs to `future`

nullnumber1 replied:
Also, what is so wrong with it? As it seemed to me, the main point of the issue you created is to decrease the startup time. I can disable AOT and revert everything back to normal variables, but it will lead to compiling at runtime, which definitely doesn't look like a fast startup.

ryukzak replied:
Got it. Don't take into account that static java classes work in different way that dynamic clojure modules.

Here I see two approach to finish this issue:

1. Add database check before actual bot startup, it is required because in other case I can restart bot and don't see misconfiguration immediately after that.
2. Move `conf` initialization inside `-main` with `bot-api`. It simplify csa module and API for bot configuration by removing magic symbols.

Selection is up to you.

nullnumber1 commented:
I picked the second approach of avoiding the magic symbols. Also made informative messages of db connection failure :full_moon_with_face:

nullnumber1 commented:
Please check my changes. Tried to do according to your ideas

ryukzak reviewed (changes requested):
Just fix CI issues.

nullnumber1 commented:
Added declaration of bot-api, id, message outside the function like it was there previously. Local run of linter was successful after that

nullnumber1 commented:
Ran `clojure -M -m cljfmt.main fix`. Now main file stands to its standards and passes fmt check. Sorry that haven't ran this command locally before
